New SFStoreRegisterGateway resource

matt6697 opened the pull request:
New SFStoreRegisterGateway to register a Netscaler Gateway with an existing Citrix StoreFront store for use when users access StoreFront from the internet.

Updated readme.md with the documentation of this new resource.

matt6697 commented:
@iainbrighton Hi Iain, wish you a happy new year full of Citrix and DSC deployments;-) Did you have some time to have a look to this new resource ? Cheers. Matthieu

iainbrighton commented on DSCResources/VE_SFStoreRegisterGateway/VE_SFStoreRegisterGateway.ps1:
```diff
@@ -0,0 +1,216 @@
+        elseif ($Ensure -eq 'Present') {
+            #The CitrixAGBasic authentication protocol is used when accessing StoreFront remotely through a Citrix NetScaler Gateway
+            $AuthenticationService = Get-STFAuthenticationService -VirtualPath $storeService.AuthenticationServiceVirtualPath;
+            [ref] $null = Add-STFAuthenticationServiceProtocol -AuthenticationService $authenticationService -Name 'CitrixAGBasic';
```
Will this throw if the 'CitrixAGBasic' method is already added?

matt6697 replied:
I confirm that it will not throw an error if the 'CitrixAGBasic' method is already present. Tested by adding a second gateway to a store already configured with a Netscaler Gateway and CitrixAgBasic authentication

iainbrighton commented on DSCResources/VE_SFStoreRegisterGateway/VE_SFStoreRegisterGateway.ps1:
```diff
@@ -0,0 +1,216 @@
+            [ref] $null = Add-STFAuthenticationServiceProtocol -AuthenticationService $authenticationService -Name 'CitrixAGBasic';
+
+            if($DefaultGateway -and $UseFullVpn) {
+                [ref] $null = Register-STFStoreGateway -Gateway $GatewayObject -StoreService $storeService -DefaultGateway -UseFullVpn; 
```
Presumably, the `Register-STFStoreGateway` calls here won't throw if the gateway is already registered? Will it only update (and the same for the next 3 calls)?

matt6697 replied:
I confirm that it will not throw an error  if the gateway is already registered.

Tested without error by updating the usefullvpn variable after gateway registration

**First call:**
```
$invokeParams = @{
  Name          = 'SFStoreRegisterGateway'
  Method        = 'set'
  Property      = @{
    storevirtualpath = '/Citrix/Store'
    gateway = 'NSGateway'
    defaultgateway = $true
    usefullvpn = $false
    ensure = 'Present'
  }

PS C:\DSC> C:\DSC\SFStoreRegisterGateway_test.ps1
{"rebootrequired":false,"indesiredstate":true,"errormessage":""}
```

**Update call:**
```
$invokeParams = @{
  Name          = 'SFStoreRegisterGateway'
  Method        = 'set'
  Property      = @{
    storevirtualpath = '/Citrix/Store'
    gateway = 'NSGateway'
    defaultgateway = $true
    usefullvpn = $true
    ensure = 'Present'
  }

PS C:\DSC> C:\DSC\SFStoreRegisterGateway_test.ps1
{"rebootrequired":false,"indesiredstate":true,"errormessage":""}
```

iainbrighton commented on README.md:
```diff
@@ -305,3 +306,34 @@ SFStoreWebReceiver [String] #ResourceName
+SFStoreRegisterGateway [String] #ResourceName
+{
+    StoreVirtualPath = [string]
+	Gateway = [string]
```
Indent/spacing appears to be out.

iainbrighton commented on README.md:
```diff
@@ -305,3 +306,34 @@ SFStoreWebReceiver [String] #ResourceName
+	Gateway = [string]
+    [ SiteId = [UInt16] ]
+    [ DefaultGateway = [bool] ]
+	[ UseFullVpn = [bool] ]
```
Indent/spacing appears to be out here too.

iainbrighton commented:
@matt6697 Looks good to me. There are a couple of comments - the most important being that I can 't see the .mof file 😢.

matt6697 commented:
@iainbrighton : Tks for the quick review ;-) I corrected indent/spacing in readme and just updated the encoding of the MOF file to UTF-8 (error was probably when editing it with powershell ise). You should be able to see it. Sorry for the inconvenience during review.

I will double check this weekend to confirm that the `Register-STFStoreGateway` function only updates the data but I think that it was the case during the tests in december. 

It won't throw if the `CitrixAGBasic` is already present as I was able to use this function to add a second Gateway to a store already configured with a first one. I will also double check this weekend.

matt6697 commented:
@iainbrighton : Double checked your `Register-STFStoreGateway` and `CitrixAGBasic` questions. Both calls don't throw an error during update if the gateway or the authentication method are already configured.
